### ISSUE TITLE
Continuous integration

### ISSUE THREAD

**Issue by shlyakpavel** [lxqt/lxqt#1644]
##### Expected Behavior
LxQT components are tested on all platforms after each PR/commit

##### Current Behavior
I do not see anything like that

##### Possible Solution
Integrate repos with Travis CI or any other service

##### Context
I guess some PRs can break build on some of supported platforms, such as *BSD or compilers.  It would be also nice to support Mac OS as most of your software does run well on it. Moreover, one could perform static analysis on each build in order to prevent regressions.
P. S. I have seen some comments on Travis here in 2016, seems like something has happened since that.

**Comment by tsujan** [comment]
> LxQT components are tested on all platforms after each PR/commit

Each LXQt dev compiles the changes he's made on his own Linux distro. Usually, other devs do the same (at least one but often all devs) and, of course, all devs will do it soon. There also users that follow the latest git (especially Arch users).

As for *BSD, its users should do (and have done)  it.

It isn't true that everything written on/for Linux can be used on Mac OS too. Mac needs its special handling.

**Comment by tsujan** [comment]
> I'm actually a Mac user. I have build LXQt components as well as FeatherPad/FeatherNotes on Mac...

Very interesting! Do they work without problem? All of them?

If you needed to make small changes for Mac, please make PRs!

> Qt is really nice for that

Yes! Sometimes it's surprising how multi-platform Qt can be. An MS Windows user once told me he successfully compiled Kvantum on Windows after small changes. Mac may be understandable but MS Windows and a style plugin?!

> And for static analysis...

PRs are welcome.

**Comment by shlyakpavel** [comment]
> Very interesting! Do they work without problem? All of them?

Most of them :)
> 
> If you needed to make small changes for Mac, please make PRs!

Ok) I'll do it a bit later

**Comment by agaida** [comment]
i have nothing against continous integration - but there are some problems:
* Henn and Egg - you need the latest components to build in a ci - otherwise the tests are useless
* travis and the supported Qt versions - there are some really fugly workarounds

The only sane solution i see is travis on a sid base with latest snapshot packages - and that is what i run most of the time. The problem with that approach is: right now i can integrate not merged PRs in these packages - that would be not so clever with an attached CI - and a second set of clean snapshots would be overkill. But the idea is nice.

**Comment by jubalh** [comment]
@shlyakpavel we have this partly covered with [OBS](https://openbuildservice.org/).

The [X11:LXQt:git](https://build.opensuse.org/project/show/X11:LXQt:git) devel project is what you need to look at.
Upon every git commit to any LXQt repo GitHub triggers a webhook, and thus notifies OBS to rebuild the package (and all it's dependencies).

So we see after each commit that LXQt still builds for openSUSE Linux. Often (most of the time) catching errors that can happen on several distros of course.

OBS can also build packages for Arch, Debian, Ubuntu, Fedora, Red Hat. But the packagers need to contribute their recipes to OBS. So far there is only the spec file for openSUSE but this already catches most of the things.

It also gives users the ability to always run the latest git packages of LXQt just by doing:

```
zypper ar -f obs://X11:LXQt:git
zypper dup --from lxqtgit
```

Then it upgrades all the packages from the normal repo to the latest git repo and from now one nobody needs to rebuild all the LXQt components himself but it will be part of normally updating your system with the package manager.

We do this since several years already.
I also tried to highlight this in the past several times but didn't get much response. Like this it is already very useful. But it would even better if we would have [openQA](http://open.qa/) tests.

So far we only have a basic tests which checks whether the latest released version of LXQt starts and ends as expected. You can see such a test temporarily here: https://openqa.opensuse.org/tests/821235

This is ran upon every openSUSE Tumbleweed snapshot.
If we would run an openQA instance we could add more tests, testing a lot of LXQt functionality (opening pcmanfm-qt, removing folders, adding them, etc) upon every commit or before a release.
But this needs time and resources. I planned to do it since 2 years but didn't find the time to do so.

**Comment by agaida** [comment]
@jubalh - right, a tool like obs is only chance that make sense. If we could setup travis on that base it would be fine.

**Comment by jubalh** [comment]
> right, a tool like obs is only thing that make sense. If we could setup travis on that base it would be fine.

I will try to find out if if's possible to do that :)

So long: at the right side of https://build.opensuse.org/project/show/X11:LXQt:git one sees that 35 packages build
Whenever there is a commit to an LXQt project master branch that does not build I get an email from OBS. Some issues I opened where discovered this way. Anybody who wants to have access to that OBS repository is welcome to message me. I can add them there too when they have an account and they would also get an email if something doesn't build.

We could (if we want) also create a branch 'development' that is not master and first push everything there, then have OBS test from that branch, this way we would have `master` always buildable and first let OBS check. But I don't think this is necessary at this point. Probably not even wise yet as we are too few people.

In any case I'm happy if some people get more interest of want to contribute, unfortunately I, like so many people, have not enough time to do all the things I would like to :-)

Here is a nice talk by Andrew Lee about OBS for people who don't know about it [Open Build Service in Debian - It is impressive how much time and resources a team can save](https://www.youtube.com/watch?v=aSMw_hvBYnw)

**Comment by agaida** [comment]
meh - the point of a ci is to check PRs before they are committed :D

**Comment by tsujan** [comment]
Warning about possible overrating: How many times did we have such PR problems?

**Comment by agaida** [comment]
@tsujan: right, the amount of false positves and the need for maintainance might be higher as the time needed to fix a wrongly committed PR

**Comment by tsujan** [comment]
I agree.

**Comment by agaida** [comment]
@jubalh - i talked sometimes about OBS with Andrew - but i see some downsides when not in OpenSuse - damn hard to set up (if using the debian packages), damn hard to set up in OpenSuse (if using the official packages)

**Comment by jubalh** [comment]
> meh - the point of a ci is to check PRs before they are committed :D

Like I said a special branch for that would be a hack around it.

> Warning about possible overrating: How many times did we have such PR problems?

Several times already.

> @jubalh - i talked sometimes about OBS with Andrew - but i see some downsides when not in OpenSuse - damn hard to set up (if using the debian packages), damn hard to set up in OpenSuse (if using the official packages)

We don't need to set up an OBS instance. We only need to add recipes to the packages in X11:LXQt:git.
So far we have rpm spec files, we only need to add the others. OBS does the rest.

Anyways we don't need to have this set up for various distro to catch most of such errors like mentioned above. Most of such errors are distro agnostic. What we need is more people taking a look at the project ;)

**Comment by yan12125** [comment]
Here's a proof-of-concept using Travis CI + Arch Linux: https://github.com/lxqt/qtermwidget/pull/271 :)

**Comment by tsimonq2** [comment]
Lubuntu has [a CI](https://ci.lubuntu.me/) with Jenkins that we build all of LXQt in. The code is a bit of a prototype, but it does the job.

**Comment by yan12125** [comment]
Looks great! Can it build pull requests?

**Comment by tsimonq2** [comment]
I don't believe it can, but it might be possible. Right now it just takes the tip of the `ci/unstable` branch of our packaging, grabs the tip of master, wraps that in a tarball, and uploads it to a PPA. The key here would be programmatically making a job that would test a specific pull request; if it's in a different branch I can switch individual builds to it, but it would replace the existing job.

It's really meant for downstream use for how it's written right now. For upstream use, it'd need some work.

**Comment by agaida** [comment]
@yan12125 - how should it work reliable - esp. if some more components are involved? This comes esp. true for our libs. Fine, they will build - but whats about things like symbol changes, needed rebuilds of other (dependend) packages and so on?

See: https://github.com/lxqt/lxqt/issues/1644#issuecomment-450642806 - so what do we get from such a ci - only the knowledge that a packge build but not anything about that it will work correctly.

**Comment by agaida** [comment]
ok - the arch approach looks promising and should work fine with our commit speed - the only downside i see are abi/api changes after our base libs are changed - so what, don't happend such often.

**Comment by yan12125** [comment]
@agaida you're right, the current setup can't handle ABI/API changes automatically. If that happens, please ping me and I'll do the rebuilds and push packages to [archlinuxcn].

> only the knowledge that a packge build but not anything about that it will work correctly.

Yes, only it builds or not. Still need to pull for functionality checks. It's possible to add functional tests in the future. Let's do some simple steps for now :)

**Comment by yan12125** [comment]
Travis CI is set up for three projects (qtermwidget, lxqt-panel, pcmanfm-qt), and they have been working fine until recently. In https://github.com/lxqt/pcmanfm-qt/pull/1271 and https://github.com/lxqt/lxqt-panel/pull/1533, tests fail due to the pull rate limit of Docker Hub. The issue is discussed at [a Travis CI forum thread](https://travis-ci.community/t/docker-hub-you-have-reached-your-pull-rate-limit/10517), but no solution is proposed yet.

Meanwhile, another CI platform, [GitHub Actions](https://github.com/features/actions) is confirmed by a staff that [they are working towards to limit](https://github.com/actions/virtual-environments/issues/1445#issuecomment-713861495). I've been using GitHub actions for a private project at my work, and the experience is slightly better than Travis CI, so I'm now considering switching to GitHub Actions. I consider it a high-priority task, so if nobody has objection, I will start the migration this weekend.

**Comment by tsujan** [comment]
> so if nobody has objection, I will start the migration this weekend.

I don't have enough knowledge on the subject, but If it's better, why not?

**Comment by yan12125** [comment]
Besides ideological reasons, the only disadvantage of GitHub Actions is viewing build logs requires logging into an GitHub account even for public repositories.

**Comment by yan12125** [comment]
@tsujan @stefonarch Could you have a look on three pull requests linked above?

**Comment by tsujan** [comment]
@yan12125
As I said above, I know almost nothing about CI. But I trust your work. Please merge them; we'll see what happens in practice.

**Comment by stefonarch** [comment]
So as me, I can see more or less what it does and it looks fine.

**Comment by yan12125** [comment]
Thanks a lot for the trust! I merge them all. Existing pull requests need be rebased against/merged with the latest git master to utilize the new CI platform.

**Comment by yan12125** [comment]
Recent CI breakages (https://github.com/lxqt/lxqt-panel/pull/1562, https://github.com/lxqt/qtermwidget/pull/404, https://github.com/lxqt/pcmanfm-qt/pull/1305) make me think about a shared script among all CI-enabled projects. My first thought is placing the shared script in lxqt-build-tools and cloning it in all other projects. Any better ideas?

**Comment by stefonarch** [comment]
Somehow strange that it will be in build-tools which many people have installed.
Maybe an extra repo for stuff like this?

**Comment by yan12125** [comment]
Sounds not bad. Will give it a try.

**Comment by yan12125** [comment]
> Maybe an extra repo for stuff like this?

Added in https://github.com/lxqt/lxqt-panel/pull/1566, https://github.com/lxqt/pcmanfm-qt/pull/1311, https://github.com/lxqt/qtermwidget/pull/405. Hopefully I don't need to create pull requests for all projects anymore with this "shared CI" strategy.

**Comment by stefonarch** [comment]
Looks like working, merged PR in panel, thanks!

**Comment by yan12125** [comment]
A CI failure is observed in https://github.com/lxqt/pcmanfm-qt/pull/1321#issuecomment-797285095. The cause is that the latest pcmanfm-qt wants the latest libfm-qt package, while the latter was not updated yet - it is updated periodically. For now I don't have a good idea to fix such issues. Please comment if anyone has an idea.

**Comment by tsujan** [comment]
@yan12125 
IMHO, that's acceptable and we don't need to worry about such cases. They're inevitable once in a while.

**Comment by yan12125** [comment]
Not a priority for me, but nice to have ;)

**Comment by yan12125** [comment]
There are issues around CI for highly related components, like libfm-qt/pcmanfm-qt and qtermwidget/qterminal. Currently, CI tests pull prebuilt -git packages from [archlinuxcn] repo for dependencies, so building against a specific branch does not work. A solution might be supporting specifying branch names for some dependencies. This requiers building dependencies from sources in CI tests as well as parsing commit messages or pull request descriptions for the branch name.

**Comment by yan12125** [comment]
There are many issues with the current CI approach. One is mentioned in my previous comment. Another is that there is always a delay between git commits and package updates in the [archlinuxcn] repo. The latter becomes worse when PKGBUILDs cannot catch up changes in git-master in time. I propose to disable all current CI scripts until there is a better approach.
